sklearn/metrics/pairwise.py

```diff
@@ -551,7 +565,254 @@ def cosine_distances(X, Y=None):
+        Indicates with True/False whether a column is a categorical attribute.
+        This is useful when categorical atributes are represented as integer
+        values. Categorical ordinal attributes are treated as numeric, and
+        must be marked as false.
+
+        Alternatively, the categorical_features array can be represented only
+        with the numerical indexes of the categorical attribtes.
+
+        If the categorical_features array is not provided, the categorical
+        columns will be auto-detected. Columns with strings, boolean,
+        and other object types will be identified as categorical, and columns
+        with any numeric type will be identified as non categorical.
+
+    apply_scaling : optional boolean, default True
+        Indicates if the data will be scaled between 0 and 1. If false, it
+        is assumed the data is already scaled.
+
+    Returns
+    -------
+    similarities : ndarray, shape (n_samples, n_samples)
+
+    References
+    ---------
+    Gower, J.C., 1971, A General Coefficient of Similarity and Some of Its
+    Properties.
+
+    Notes
+    -----
+    The numeric feature ranges are determined from both X and Y.
+
+    No support for sparse matrices.
+
+    The object type and its subtypes (e.g., str), are treated as categorical
+    features.
+
+    This implementation modifies the Gower's original similarity measure in the
+    folowing aspects:
+
+    * The values in the original similarity S range between 0 and 1. To
+    guarantee this, it is assumed the numerical features of X and Y are
+    scaled between 0 and 1.
+
+    * Different from the original similarity S, this implementation returns
+    1-S.
+
+    """
+    if issparse(X) or issparse(Y):
+        raise TypeError("Sparse matrices are not supported for gower distance")
+
+    if X is None or len(X) == 0:
+        raise ValueError("X can not be None or empty")
+
+    y_none = Y is None
+
+    # It is necessary to convert to ndarray in advance to define the dtype
+    # as np.object, otherwise numeric columns will be converted to string
+    # if there are other string columns.
+    if not isinstance(X, np.ndarray):
+        X = np.asarray(X, dtype=np.object)
+
+    if not y_none and not isinstance(Y, np.ndarray):
+        Y = np.asarray(Y, dtype=np.object)
+
+    X, Y = check_pairwise_arrays(X, Y, precomputed=False, dtype=X.dtype,
+                                 force_all_finite=False)
+
+    n_rows, n_cols = X.shape
+
+    categorical_features_obj, categorical_features_num = \
+        _detect_categorical_features(X, categorical_features)
+
+    # Categorical columns
+    X_cat_obj = X[:, categorical_features_obj].astype(np.object)
+    X_cat_num = X[:, categorical_features_num].astype(np.float32)
+
+    # Numerical columns
+    X_num = X[:, np.logical_not(categorical_features_obj |
+                                categorical_features_num)].astype(np.float32)
```

Comment:
I'm not sure about this forced float32 cast.

Reply:
You can use our `utils.validation.check_array` for this purpose, which can change the dtype only if it's not already numerical.